### ISSUE TITLE
Package 'Microsoft.ApplicationInsights.AzureWebSites' installing failing with ASP.NET project

### ISSUE THREAD

**Issue by jainaashish** [NuGet/Home#6892]
Steps to repro:
1. Create ASP.NET web application (.NET framework)
2. Try installing 'Microsoft.ApplicationInsights.AzureWebSites' version 2.5.6
3. It will fail with following error:
```
Error downloading 'Microsoft.ApplicationInsights.AzureWebSites.2.5.6' from 'https://api.nuget.org/v3-flatcontainer/microsoft.applicationinsights.azurewebsites/2.5.6/microsoft.applicationinsights.azurewebsites.2.5.6.nupkg'.
The specified path, file name, or both are too long. The fully qualified file name must be less than 260 characters, and the directory name must be less than 248 characters.
```

Another case, if you try "dotnet add package <this package name>" then it works fine and is able to get this package on disk in global packages folder.

So there are multiple issues here:
1. Package signing verification is failing as part of VS which throws this error so needs better error handing
2. Package signing verification succeed with dotnet add package so which indicates there might be different verification behavior... needs to confirm.
3. Package itself is very weird, lot of data have been just dumped into `/content` folder so should discuss the purpose of this package with the authors.
4. PackageType for this package is defined as "AzureSiteExtension" which is again wrong.

Work with AppInsight team to fix the package as well as error handing in NuGet.

**Comment by mishra14** [comment]
We should remove the signing tag because - 
1. @zhili1208 confirmed that signing verification succeeds in VS
2. dotnet add does not run signing verification. That is cross platform work that is yet to be done.

Based on Zhi's response the issue is a long file path problem during extraction.

**Comment by nkolev92** [comment]
@zhili1208 Can you add some details from the investigation here?
It's boils down to a long file path problem.

**Comment by rrelyea** [comment]
Zhi - as we mentioned in person, please work with this package author to help them understand how to better support users.
Also, please go write a validation rule that helps warn package authors when their packages longfilepath is too long. (we need to decide what too long is...) This should be one of the validations we run in pack, and in nuget.org ingestion (eventually) -- talk with @rohit21agrawal & @nkolev92 about those validation rules.

**Comment by jainaashish** [comment]
@nkolev92 dupe it out to search by tfm issue

**Comment by nkolev92** [comment]
Duplicate of #4071.

**Comment by gckeller-ps** [comment]
I have my project in the drive root and still get this exception.  I can't find any way to install it.  Suggestions?

**Comment by cijothomas** [comment]
@gckeller-ps Microsoft.ApplicationInsights.AzureWebSites is not meant to enable monitoring from visual studio. Its a site extension for Azure App Service, and has no value installing from visual studio.

To enable application insights from Visual Studio, please install Microsoft.ApplicationInsights.Web - this is the right package.

**Comment by gckeller-ps** [comment]
@cijothomas This is not accurate.  Microsoft.ApplicationInsights.Web is for .NET Framework 4.5: https://www.nuget.org/packages/Microsoft.ApplicationInsights.Web
The extension you are referring to is for profiling which we are not using for profiling as .NET can only have 1 profiler and we are using New Relic for profiling: https://docs.newrelic.com/docs/agents/net-agent/azure-troubleshooting/net-profiler-conflict-microsoft-application-insights
I could be wrong, but I don't believe I added the Microsoft.ApplicationInsights.AzureWebSites package as it's not included in the branch that I've been doing the AI and New Relic work in and I don't believe that I've done any merging into develop.  

If you want VS profiling, please work with Eric to enter a JIRA ticket for the request and I'll get 'er done.

There were a bunch of issues with both AI and New Relic in develop and I have corrected all of that in GLACIERMVP-173.
Application Insights is not just for Monitoring, but has a number of features and functionality.  The packages that I'm using are as follows:
Microsoft.ApplicationInsights.AspNetCore
Microsoft.Extensions.Logging.ApplicationInsights
Both of these packages are required for what we are using AI for.  I have tested these and they are working as expected in this feature branch.
I have not merged into develop yet as I'm waiting on a decision on the Windows or Linux/Docker front to include the proper New Relic library.  The official assessment document is in Eric's hands right now, but below is a summary list of limitations in the Linux/Docker world that will likely pose a number of headaches, roadblocks and limitations in the future:

Linux/Docker Specific Challenges
•	Does not support .NET Core 2.2; Docker image is required to run 2.2 which adds overhead.  
•	Continued security vulnerabilities found in container-based solutions such as Docker and Kubernetes: https://www.zdnet.com/article/doomsday-docker-security-hole-uncovered/ 
•	Limited New Relic support.  The following are NOT supported on .NET Core Agent installed on Linux
o	Thread profiling on Linux (thread profiling on Windows is supported)
o	Memory usage on Linux due to an ongoing .NET Core issue 
o	Automatic New Relic Browser script injection (API or manual instrumentation is required)
•	Microsoft releases new features on Windows first and it takes significantly longer to become available on Linux (if at all)
•	Microsoft does not directly support Linux.  Microsoft relies on their distribution partners for support and each one offers various levels of support.
•	Third-party open source support is limited to normal business hours and is not 24x7
•	Application Insights support is poor and Application Insights is a foundational aspect of Azure that spans a number of different features and functionality (Please see provided App Insights Appendix)
•	Does not support virtual network integration
•	Does not support Azure Active Directory/third-party authentication
•	Does not support Kudu site extensions
•	Does not support Run from Package
•	Does not support Process Explorer
•	Does not support App Cloning
•	Does not support App Service Editor
•	Does not support Web Jobs
•	Does not support mobile push notifications
•	Does not support more than 1 port
•	Does not support changing App Service Plans
•	You cannot create Web App for Containers in an App Service plan already hosting non-Linux Web Apps.
•	“based on a current limitation, don't mix Windows and Linux apps in the same resource group.” – Microsoft; this seems like an odd recommendation, but it is from Microsoft
•	Auto-scaling can be integrated with Application Insights metrics.  It is unknown how greatly auto-scaling metrics are impacted by AI/Linux limitations, but because New Relic is the preferred profiling provider and New Relic and App Insights profiler cannot be run side-by-side (https://docs.newrelic.com/docs/agents/net-agent/azure-troubleshooting/net-profiler-conflict-microsoft-application-insights), it is presumed that custom metrics cannot be used for auto-scaling as described in the auto-scaling best practices document.
References
•	https://docs.microsoft.com/en-us/azure/app-service/containers/app-service-linux-intro
•	https://docs.microsoft.com/en-us/azure/app-service/containers/app-service-linux-faq
•	https://www.zdnet.com/article/doomsday-docker-security-hole-uncovered/
•	https://docs.newrelic.com/docs/agents/net-agent/getting-started/compatibility-requirements-net-core-agent 
•	https://docs.microsoft.com/en-us/azure/architecture/best-practices/auto-scaling
•	https://blog.kloud.com.au/2018/11/06/deploying-app-services-with-run-from-package-azure-storage-and-azure-pipelines/ 
•	https://docs.newrelic.com/docs/agents/net-agent/azure-troubleshooting/net-profiler-conflict-microsoft-application-insights
•	https://blogs.msdn.microsoft.com/waws/2017/09/08/things-you-should-know-web-apps-and-linux/ 
•	https://docs.microsoft.com/en-us/azure/architecture/reference-architectures/app-service-web-app/multi-region
•	https://support.microsoft.com/en-us/help/2941892/support-for-linux-and-open-source-technology-in-azure
•	https://docs.microsoft.com/en-us/azure/virtual-machines/linux/endorsed-distros
•	https://blogs.msdn.microsoft.com/linuxonazure/
•	https://blogs.msdn.microsoft.com/linuxonazure/2016/08/07/best-practices-running-linux-on-azure/
•	https://blogs.msdn.microsoft.com/appserviceteam/2018/07/30/linuxasega/
•	https://www.petri.com/azure-app-services-linux 
•	http://itnerd.space/2016/11/02/azure-app-service-architecture-3-app-service-on-linux/

**Comment by gckeller-ps** [comment]
Thanks for you feedback @cijothomas  Sorry, I responded previously believing this was an internal comment.  I actually really dug into this one and found the problem last week.  I was not looking to do VS in app insights.  This package is actually screwed up.  I used the process monitors to identifying the problematic files and observing the file system activity.  They were numerous.  If you open the actual nuget and explore the contents the paths within the nuget actually go beyond the maximum length before even making it to your local drive.  Really difficult to identify without using process explorer.  The only way I could get this to work was thru App Service Site Extensions.  It is not possible to add a reference to this library at all.  The package needs to be fixed or Nuget needs to support long filenames.  SPECIFICALLY, I did need to use this when deploying to Linux (.NET Core) as Linux doesn't support site extensions.  This is the only way.  I'm hoping that I can convince the client to go with a Windows deployment as they will not be able to use App Insights without this package.  BTW, this is a very large client having this problem.  2K hits/second.  Microsoft should view this as a priority as they may lose a big customer using app insights.  This package is a required dependency if you are using Core and deploying on Linux.

**Comment by gckeller-ps** [comment]
This ticket should also be re-opened now that I've identified the problem.  It is a problem with the package and not Nuget and that it actually precludes the use of App Insights if you're deploying .NET Core to Linux.

**Comment by cijothomas** [comment]
@gckeller-ps For .net core, we have a package Microsoft.ApplicationInsights.AspNetCore which can be installed via nuget from visual studio and enable application insights in code. This should allow usage of application insights in linux.
Microsoft.ApplicationInsights.AzureWebSites is not going to be supported in future.

**Comment by gckeller-ps** [comment]
That is correct; however, once deployed it throws an exception on startup that requires a class in Microsoft.ApplicationInsights.AzureWebSites.  It has to do with the .NET Core bootstrapping of the Microsoft.ApplicationInsights.AspNetCore library.  This problem also occurs on windows, but I can confirm that it can be resolved by installing the site extension.  It cannot be resolved on Linux because Linux does not support site extensions.

**Comment by cijothomas** [comment]
@gckeller-ps https://github.com/Microsoft/ApplicationInsights-aspnetcore/issues - Pleaese an issue in this repo with the issue you are having with linux+ .net core. We can definitely investigate more there.
